portfolio/src/main/java/com/google/sps/servlets/MarkersServlet.java

```diff
@@ -0,0 +1,127 @@
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("application/json");
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query("Marker");
+
+        ArrayList<Marker> markers = new ArrayList<>();
+
+        PreparedQuery results = datastore.prepare(query);
+
+        for (Entity entity : results.asIterable()) {
+        long id = entity.getKey().getId();
+        double lat = (double) entity.getProperty("lat");
+        double lng = (double) entity.getProperty("lng");
+        String userId = (String) entity.getProperty("userId");
+
+        Marker marker = new Marker(lat, lng, userId, id);
+        markers.add(marker);
+        }
+
+        Gson gson = new Gson();
+
+        response.getWriter().println(gson.toJson(markers));
+
+    }
+
+    /** Receives and stores a new marker, then sends it. */
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        UserService userService = UserServiceFactory.getUserService();
+
+        String userId = userService.getCurrentUser().getUserId();
+        double lat = Double.parseDouble(request.getParameter("lat"));
+        double lng = Double.parseDouble(request.getParameter("lng"));
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+        Entity entity = new Entity("Marker");
+        entity.setProperty("lat", lat);
+        entity.setProperty("lng", lng);
+        entity.setProperty("userId", userId);
+
+        datastore.put(entity);
+
+        long id = entity.getKey().getId();
+
+        Marker marker = new Marker(lat, lng, userId, id);
+
+        Gson gson = new Gson();
+
+        response.getWriter().println(gson.toJson(marker));
+
+    }
+
+    /** Deletes Markers */
+    @Override
+    public void doDelete(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        
+        UserService userService = UserServiceFactory.getUserService();
+
+        String userId = userService.getCurrentUser().getUserId();
+
+        double lat = Double.parseDouble(request.getParameter("lat"));
+        double lng = Double.parseDouble(request.getParameter("lng"));
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+        Query query = new Query("Marker");
+        Query.CompositeFilter queryFilter = new Query.CompositeFilter(Query.CompositeFilterOperator.AND, Arrays
+                .asList(new Query.FilterPredicate("userId", Query.FilterOperator.EQUAL, userId), new Query
+                .FilterPredicate("lat", Query.FilterOperator.EQUAL, lat), new Query
+                .FilterPredicate("lng", Query.FilterOperator.EQUAL, lng)));
+
+        query.setFilter(queryFilter);
+
+        PreparedQuery results = datastore.prepare(query);
+        Entity entity = results.asSingleEntity();
```

Comment:
Is this one Entity actually a list of Marker entities?

Reply:
This should be a single marker, but now that you mention it there might exist a chance of the same user creating two exact markers so this would result in multiple entities within this query. I'll change this so it uses the unique id instead of lat and lng to avoid this scenario!

Reply:
Sounds good!